doc/source/ray-air/examples/torch_detection.md

```diff
@@ -0,0 +1,476 @@
+
+
+class VOCAnnotationDatasource(FileBasedDatasource):
+    def _read_file(self, f: pa.NativeFile, path: str, **reader_args) -> pd.DataFrame:
+        text = f.readall().decode("utf-8")
+        annotation = xmltodict.parse(text)["annotation"]
+
+        objects = annotation["object"]
+        # If there's one object, `objects` is a `dict`; otherwise, it's a `list[dict]`.
+        if isinstance(objects, dict):
+            objects = [objects]
+
+        boxes: List[Tuple] = []
+        for obj in objects:
+            x1 = float(obj["bndbox"]["xmin"])
+            y1 = float(obj["bndbox"]["ymin"])
+            x2 = float(obj["bndbox"]["xmax"])
+            y2 = float(obj["bndbox"]["ymax"])
+            boxes.append((x1, y1, x2, y2))
+
+        labels: List[int] = [CLASS_TO_LABEL[obj["name"]] for obj in objects]
+
+        filename = annotation["filename"]
+
+        return pd.DataFrame(
+            {
+                "boxes": TensorArray([boxes]),
+                "labels": TensorArray([labels]),
+                "filename": [filename],
+            }
+        )
+
+    def _rows_per_file(self):
+        return 1
+```
+
+### Read annotations
+
+To load the annotations into a `Dataset`, call `ray.data.read_datasource` and pass
+the custom datasource to the constructor. Ray will read the annotations in parallel.
+
+```python
+import os
+import ray
+
+
+root = os.path.abspath("./VOCdevkit/VOC2012")
+annotations: ray.data.Dataset = ray.data.read_datasource(
+    VOCAnnotationDatasource(), paths=os.path.join(root, "Annotations")
+)
+```
+
+Look at the first two samples. `VOCAnnotationDatasource` should've correctly parsed
+labels and bounding boxes.
+
+```python
+annotations.take(2)
+```
+
+### Load images into memory
+
+
+Each row of `annotations` contains the filename of an image.
+
+Write a user-defined that loads these images. For each annotation, your function will:
+1. Open the image associated with the annotation.
+2. Add the image to a new `"image"` column.
+
+```python
+from typing import Dict
+
+import numpy as np
+from PIL import Image
+
+
+def read_images(batch: Dict[str, np.ndarray]) -> Dict[str, np.ndarray]:
+    images: List[np.ndarray] = []
+    for filename in batch["filename"]:
+        path = os.path.join(root, "JPEGImages", filename)
+        image = np.array(Image.open(path))
+        images.append(image)
+    batch["image"] = np.array(images, dtype=object)
+    return batch
+
+
+dataset = annotations.map_batches(read_images)
+dataset
+```
+
+### Split the dataset into train and test sets
+
+
+Once you've created a `Dataset`, split the dataset into train and test sets.
+
+```python
+train_dataset, test_dataset = dataset.train_test_split(0.2)
+```
+
+## Define preprocessing logic
+
+
+A `Preprocessor` is an object that defines preprocessing logic. It's the standard way
+to preprocess data with Ray.
+
+To preprocess the images, create a `TorchVisionPreprocessor` and call
+`Preprocessor.transform` on the dataset.
```

Comment:
add more info on preprocessor vs. per-epoch preprocessor?

"We define two preprocessors: one to transpoe, normalize, and convert all the images to a torch tensor (`transforms.ToTensor()`), and another one that will apply a random flip each epoch. The randomness will make our model more robust. We can pass both of these preprocessors into a Trainer."

Reply:
also we're no longer directly calling transform right?